File: 回測模組/exporters/early_chart_creator.py
"""
早盤策略專用圖表產生器
只顯示關鍵進場條件：大量搓合金額 + 五檔掛單比率
"""

import logging
import pandas as pd
import plotly.graph_objects as go
from plotly.subplots import make_subplots
from typing import Dict, List, Optional

logger = logging.getLogger(__name__)


class EarlyChartCreator:
    """早盤策略圖表建立器（簡化版）"""

    def __init__(self, config: Optional[Dict] = None):
        """
        初始化圖表建立器

        Args:
            config: 設定字典
        """
        self.config = config or {}

    def create_chart(
        self,
        df: pd.DataFrame,
        trade_record: Dict,
        output_path: str,
        stock_id: str,
        date: str,
        massive_threshold: float
    ) -> str:
        """
        建立早盤策略視覺化圖表（簡化版）

        Args:
            df: 資料 DataFrame
            trade_record: 交易記錄
            output_path: HTML 輸出路徑
            stock_id: 股票代碼
            date: 日期
            massive_threshold: 大量搓合門檻（元）

        Returns:
            輸出路徑
        """
        # 過濾有效價格資料
        df_valid = df[df['price'] > 0].copy() if 'price' in df.columns else df.copy()

        if len(df_valid) < 10:
            logger.warning(f"{stock_id} 有效資料過少，跳過圖表產生")
            return None

        # 建立子圖（3行1列）
        fig = make_subplots(
            rows=3, cols=1,
            shared_xaxes=True,
            vertical_spacing=0.05,
            row_heights=[0.50, 0.25, 0.25],
            subplot_titles=(
                f'{stock_id} 早盤進場策略 (09:01-09:05) - {date}',
                '大量搓合金額 (萬元)',
                '五檔掛單比率 (委賣/委買)'
            )
        )

        # 子圖1：價格 + Day High + 進出場點 + 移動停利低點
        self._add_price_chart(fig, df_valid, trade_record, row=1, col=1)

        # 子圖2：大量搓合金額
        self._add_massive_matching_chart(fig, df, massive_threshold, row=2, col=1)

        # 子圖3：五檔掛單比率
        self._add_orderbook_ratio_chart(fig, df, row=3, col=1)

        # 更新佈局
        self._update_layout(fig, stock_id, date)

        # 儲存圖表
        fig.write_html(
            output_path,
            config={'displayModeBar': True, 'scrollZoom': True}
        )
        logger.info(f"圖表已儲存: {output_path}")

        return output_path

    def _add_price_chart(self, fig, df, trade_record, row, col):
        """添加價格圖表"""
        # 計算 Y 軸範圍
        price_min = df['price'].min()
        price_max = df['price'].max()

        if 'day_high' in df.columns:
            price_max = max(price_max, df['day_high'].max())

        price_range = price_max - price_min
        y_min = max(0, price_min - price_range * 0.05)
        y_max = price_max + price_range * 0.05

        # 價格線
        fig.add_trace(
            go.Scatter(
                x=df['time'], y=df['price'],
                mode='lines',
                name='價格',
                line=dict(color='black', width=1.5),
                hovertemplate='價格: %{y:.2f}<br>時間: %{x}'
            ),
            row=row, col=col
        )

        # Day High 線
        if 'day_high' in df.columns:
            fig.add_trace(
                go.Scatter(
                    x=df['time'], y=df['day_high'],
                    mode='lines',
                    name='Day High',
                    line=dict(color='red', width=2, dash='solid'),
                    opacity=0.7,
                    hovertemplate='Day High: %{y:.2f}<br>時間: %{x}'
                ),
                row=row, col=col
            )

        # 移動停利低點線
        colors = {'low_1m': 'green', 'low_3m': 'orange', 'low_5m': 'purple'}
        names = {'low_1m': '1分鐘低點', 'low_3m': '3分鐘低點', 'low_5m': '5分鐘低點'}

        for col_name, color in colors.items():
            if col_name in df.columns:
                fig.add_trace(
                    go.Scatter(
                        x=df['time'], y=df[col_name],
                        mode='lines',
                        name=names[col_name],
                        line=dict(color=color, width=1, dash='dash'),
                        opacity=0.6,
                        hovertemplate=f'{names[col_name]}: %{{y:.2f}}<br>時間: %{{x}}'
                    ),
                    row=row, col=col
                )

        # 設定 Y 軸範圍
        fig.update_yaxes(range=[y_min, y_max], row=row, col=col)

        # 標記進場點
        entry_price = trade_record.get('entry_price', 0)
        entry_time = trade_record.get('entry_time', '')
        entry_ratio = trade_record.get('entry_ratio', 0)

        if entry_price > 0:
            fig.add_trace(
                go.Scatter(
                    x=[entry_time], y=[entry_price],
                    mode='markers+text',
                    name='進場',
                    marker=dict(color='red', size=20, symbol='circle'),
                    text=['進場'],
                    textposition='top center',
                    hovertemplate=f"進場<br>價格: {entry_price:.2f}<br>Ratio: {entry_ratio:.1f}<br>時間: %{{x}}"
                ),
                row=row, col=col
            )

        # 標記出場點（移動停利三批出場）
        exit_colors = {'1min': 'green', '3min': 'orange', '5min': 'purple'}

        if 'exits' in trade_record:
            for i, exit_info in enumerate(trade_record['exits']):
                exit_time = exit_info.get('time', '')
                exit_price = exit_info.get('price', 0)
                exit_reason = exit_info.get('reason', '')
                exit_ratio = exit_info.get('ratio', 0)

                # 判斷出場類型
                level_name = None
                for key in ['1min', '3min', '5min']:
                    if key in exit_reason:
                        level_name = key
                        break

                exit_color = exit_colors.get(level_name, 'gray')
                exit_label = f'{level_name}停利' if level_name else '出場'

                fig.add_trace(
                    go.Scatter(
                        x=[exit_time], y=[exit_price],
                        mode='markers',
                        name=exit_label if i == 0 else None,
                        showlegend=(i == 0),
                        marker=dict(color=exit_color, size=18, symbol='triangle-down'),
                        hovertemplate=f"{exit_label}<br>價格: {exit_price:.2f}<br>比例: {exit_ratio:.1%}<br>時間: %{{x}}"
                    ),
                    row=row, col=col
                )

    def _add_massive_matching_chart(self, fig, df, threshold, row, col):
        """添加大量搓合金額圖表"""
        # 檢查欄位
        amount_col = None
        for col_name in ['massive_matching_amount', 'outside_amount_1s', 'outside_1s']:
            if col_name in df.columns:
                amount_col = col_name
                break

        if amount_col is None:
            logger.warning("找不到大量搓合金額欄位")
            return

        # 轉換為萬元
        amount_data = df[amount_col] / 10000
        threshold_wan = threshold / 10000

        # 大量搓合金額線
        fig.add_trace(
            go.Scatter(
                x=df['time'], y=amount_data,
                mode='lines',
                name='外盤金額',
                line=dict(color='blue', width=2),
                hovertemplate='外盤金額: %{y:.1f}萬<br>時間: %{x}'
            ),
            row=row, col=col
        )

        # 門檻線
        fig.add_hline(
            y=threshold_wan,
            line_dash="dash",
            line_color="red",
            line_width=2,
            annotation_text=f"門檻: {threshold_wan:.1f}萬",
            annotation_position="right",
            row=row, col=col
        )

        # 可進場區域（金額 >= 門檻）
        amount_high = amount_data.copy()
        amount_high[amount_high < threshold_wan] = None

        fig.add_trace(
            go.Scatter(
                x=df['time'], y=amount_high,
                fill='tozeroy',
                mode='none',
                name='符合門檻',
                fillcolor='rgba(0, 255, 0, 0.2)',
                showlegend=True
            ),
            row=row, col=col
        )

    def _add_orderbook_ratio_chart(self, fig, df, row, col):
        """添加五檔掛單比率圖表"""
        # 檢查欄位
        ratio_col = None
        for col_name in ['bid_ask_ratio', 'orderbook_ratio', 'ask_bid_ratio']:
            if col_name in df.columns:
                ratio_col = col_name
                break

        if ratio_col is None:
            logger.warning("找不到五檔掛單比率欄位")
            return

        # 掛單比率線
        fig.add_trace(
            go.Scatter(
                x=df['time'], y=df[ratio_col],
                mode='lines',
                name='委賣/委買',
                line=dict(color='purple', width=2),
                hovertemplate='委賣/委買: %{y:.2f}<br>時間: %{x}'
            ),
            row=row, col=col
        )

        # 門檻線（1.0）
        threshold = self.config.get('orderbook_bid_ask_ratio_min', 1.0)
        fig.add_hline(
            y=threshold,
            line_dash="dash",
            line_color="red",
            line_width=2,
            annotation_text=f"門檻: {threshold:.1f}",
            annotation_position="right",
            row=row, col=col
        )

        # 可進場區域（比率 >= 門檻）
        ratio_high = df[ratio_col].copy()
        ratio_high[ratio_high < threshold] = None

        fig.add_trace(
            go.Scatter(
                x=df['time'], y=ratio_high,
                fill='tozeroy',
                mode='none',
                name='符合門檻',
                fillcolor='rgba(0, 255, 0, 0.2)',
                showlegend=True
            ),
            row=row, col=col
        )

    def _update_layout(self, fig, stock_id, date):
        """更新圖表佈局"""
        fig.update_layout(
            title_text=f"{stock_id} 早盤進場策略 (09:01-09:05) - {date}",
            title_font_size=18,
            showlegend=True,
            height=900,
            hovermode='x unified',
            plot_bgcolor='white',
            paper_bgcolor='white'
        )

        # X 軸設定
        fig.update_xaxes(title_text="時間", row=3, col=1, showticklabels=True)
        fig.update_xaxes(showticklabels=False, row=1, col=1)
        fig.update_xaxes(showticklabels=False, row=2, col=1)
        fig.update_xaxes(showgrid=True, gridwidth=1, gridcolor='lightgray')

        # Y 軸設定
        fig.update_yaxes(title_text="價格", row=1, col=1)
        fig.update_yaxes(title_text="金額(萬)", row=2, col=1)
        fig.update_yaxes(title_text="比率", row=3, col=1)
        fig.update_yaxes(showgrid=True, gridwidth=1, gridcolor='lightgray')

File: CSharp/BacktestModule/Exporters/EarlyChartCreator.cs
using System;
using BacktestModule.Core.Models;
using System.Collections.Generic;
using BacktestModule.Core.Models;
using System.Globalization;
using BacktestModule.Core.Models;
using System.IO;
using BacktestModule.Core.Models;
using System.Linq;
using BacktestModule.Core.Models;
using System.Text;
using BacktestModule.Core.Models;
using Microsoft.Extensions.Logging;
using BacktestModule.Core;
using BacktestModule.Strategy;

namespace BacktestModule.Exporters
{
    /// <summary>
    /// Early-morning strategy chart creator (simplified version).
    /// Produces an interactive HTML chart with 3 subplots:
    /// 1. Price + Day High + entry/exit markers + trailing stop low lines
    /// 2. Massive matching amount
    /// 3. Order book ratio (ask/bid)
    ///
    /// Uses embedded Plotly.js for interactive HTML output.
    /// </summary>
    public class EarlyChartCreator
    {
        private readonly Dictionary<string, object> _config;
        private readonly ILogger<EarlyChartCreator> _logger;

        /// <summary>
        /// Initializes the early chart creator.
        /// </summary>
        /// <param name="config">Configuration dictionary. Can be null.</param>
        /// <param name="logger">Logger instance. Can be null.</param>
        public EarlyChartCreator(
            Dictionary<string, object> config = null,
            ILogger<EarlyChartCreator> logger = null)
        {
            _config = config ?? new Dictionary<string, object>();
            _logger = logger ?? Microsoft.Extensions.Logging.Abstractions.NullLogger<EarlyChartCreator>.Instance;
        }

        /// <summary>
        /// Creates the early strategy visualization chart (simplified version).
        /// </summary>
        /// <param name="data">Tick data list.</param>
        /// <param name="tradeRecord">Trade record dictionary with entry/exit information.</param>
        /// <param name="outputPath">HTML output file path.</param>
        /// <param name="stockId">Stock code.</param>
        /// <param name="date">Date string.</param>
        /// <param name="massiveThreshold">Massive matching threshold in TWD.</param>
        /// <returns>Output path, or null if insufficient data.</returns>
        public string CreateChart(
            List<TickData> data,
            Dictionary<string, object> tradeRecord,
            string outputPath,
            string stockId,
            string date,
            double massiveThreshold)
        {
            // Filter valid price data
            var validData = data.Where(d => d.Price > 0).ToList();

            if (validData.Count < 10)
            {
                _logger.LogWarning("{StockId} has too few valid data points, skipping chart generation.", stockId);
                return null;
            }

            var sb = new StringBuilder();
            sb.AppendLine("<!DOCTYPE html>");
            sb.AppendLine("<html><head>");
            sb.AppendLine("<meta charset='utf-8'>");
            sb.AppendLine($"<title>{stockId} Early Entry Strategy - {date}</title>");
            sb.AppendLine("<script src='https://cdn.plot.ly/plotly-latest.min.js'></script>");
            sb.AppendLine("</head><body>");
            sb.AppendLine($"<div id='chart' style='width:100%;height:900px;'></div>");
            sb.AppendLine("<script>");

            // Build trace data
            BuildPlotlyScript(sb, data, validData, tradeRecord, stockId, date, massiveThreshold);

            sb.AppendLine("</script>");
            sb.AppendLine("</body></html>");

            // Ensure output directory exists
            string dir = Path.GetDirectoryName(outputPath);
            if (!string.IsNullOrEmpty(dir))
                Directory.CreateDirectory(dir);

            File.WriteAllText(outputPath, sb.ToString(), Encoding.UTF8);
            _logger.LogInformation("Chart saved: {OutputPath}", outputPath);

            return outputPath;
        }

        /// <summary>
        /// Builds the Plotly.js script for the 3-subplot chart.
        /// </summary>
        private void BuildPlotlyScript(
            StringBuilder sb,
            List<TickData> allData,
            List<TickData> validData,
            Dictionary<string, object> tradeRecord,
            string stockId, string date,
            double massiveThreshold)
        {
            // --- Subplot 1: Price Chart ---
            var times = validData.Select(d => $"'{d.Time:yyyy-MM-dd HH:mm:ss.fff}'").ToList();
            var prices = validData.Select(d => d.Price.ToString("F2", CultureInfo.InvariantCulture)).ToList();
            var dayHighs = validData.Select(d => d.DayHigh.ToString("F2", CultureInfo.InvariantCulture)).ToList();

            // Price line
            sb.AppendLine($"var trace_price = {{x:[{string.Join(",", times)}], y:[{string.Join(",", prices)}], mode:'lines', name:'Price', line:{{color:'black',width:1.5}}, xaxis:'x', yaxis:'y'}};");

            // Day High line
            sb.AppendLine($"var trace_dayhigh = {{x:[{string.Join(",", times)}], y:[{string.Join(",", dayHighs)}], mode:'lines', name:'Day High', line:{{color:'red',width:2}}, opacity:0.7, xaxis:'x', yaxis:'y'}};");

            // Trailing stop low lines
            var lowLines = new[] {
                ("low_1m", "green", "1min Low"),
                ("low_3m", "orange", "3min Low"),
                ("low_5m", "purple", "5min Low")
            };
            int lowTraceIdx = 0;
            foreach (var (field, color, name) in lowLines)
            {
                var lowValues = validData.Select(d => field switch
                {
                    "low_1m" => d.Low1m,
                    "low_3m" => d.Low3m,
                    "low_5m" => d.Low5m,
                    _ => 0.0
                }).ToList();

                if (lowValues.Any(v => v > 0))
                {
                    var lowStrs = lowValues.Select(v => v > 0
                        ? v.ToString("F2", CultureInfo.InvariantCulture) : "null").ToList();
                    sb.AppendLine($"var trace_low{lowTraceIdx} = {{x:[{string.Join(",", times)}], y:[{string.Join(",", lowStrs)}], mode:'lines', name:'{name}', line:{{color:'{color}',width:1,dash:'dash'}}, opacity:0.6, xaxis:'x', yaxis:'y'}};");
                }
                else
                {
                    sb.AppendLine($"var trace_low{lowTraceIdx} = {{x:[], y:[], mode:'lines', name:'{name}', xaxis:'x', yaxis:'y'}};");
                }
                lowTraceIdx++;
            }

            // Entry marker
            sb.Append("var trace_entry = {x:[], y:[], mode:'markers+text', name:'Entry', marker:{color:'red',size:20,symbol:'circle'}, text:['Entry'], textposition:'top center', xaxis:'x', yaxis:'y'};");
            if (tradeRecord.ContainsKey("entry_price") && tradeRecord.ContainsKey("entry_time"))
            {
                double entryPrice = Convert.ToDouble(tradeRecord["entry_price"]);
                if (entryPrice > 0)
                {
                    string entryTime = FormatTime(tradeRecord["entry_time"]);
                    sb.AppendLine($"trace_entry.x = ['{entryTime}']; trace_entry.y = [{entryPrice.ToString("F2", CultureInfo.InvariantCulture)}];");
                }
            }

            // Exit markers
            sb.AppendLine("var trace_exits = [];");
            if (tradeRecord.ContainsKey("exits") && tradeRecord["exits"] is List<Dictionary<string, object>> exits)
            {
                var exitColors = new Dictionary<string, string> { ["1min"] = "green", ["3min"] = "orange", ["5min"] = "purple" };
                for (int i = 0; i < exits.Count; i++)
                {
                    var exitInfo = exits[i];
                    string exitTime = exitInfo.ContainsKey("time") ? FormatTime(exitInfo["time"]) : "";
                    double exitPrice = exitInfo.ContainsKey("price") ? Convert.ToDouble(exitInfo["price"]) : 0;
                    string exitReason = exitInfo.ContainsKey("reason") ? exitInfo["reason"]?.ToString() : "";

                    string levelName = null;
                    foreach (var key in new[] { "1min", "3min", "5min" })
                    {
                        if (exitReason.Contains(key)) { levelName = key; break; }
                    }
                    string color = levelName != null && exitColors.ContainsKey(levelName) ? exitColors[levelName] : "gray";
                    string label = levelName != null ? $"{levelName} trailing stop" : "Exit";

                    sb.AppendLine($"trace_exits.push({{x:['{exitTime}'], y:[{exitPrice.ToString("F2", CultureInfo.InvariantCulture)}], mode:'markers', name:'{label}', showlegend:{(i == 0 ? "true" : "false")}, marker:{{color:'{color}',size:18,symbol:'triangle-down'}}, xaxis:'x', yaxis:'y'}});");
                }
            }

            // --- Subplot 2: Massive Matching Amount ---
            var allTimes = allData.Select(d => $"'{d.Time:yyyy-MM-dd HH:mm:ss.fff}'").ToList();

            // Use DayHighGrowthRate as a placeholder -- in real usage, the massive matching amount
            // would be a separate computed metric. Here we expose the metric columns from TickData.
            // Since the Python code reads 'massive_matching_amount' or 'outside_amount_1s' columns,
            // we use what is available. In the C# TickData, these computed metrics are stored during the loop.
            // For this chart, we will output zero if the data is not populated.
            var amounts = allData.Select(d =>
            {
                // Use a generic field; in production, the backtest loop populates a MassiveMatchingAmount field
                // For now use 0 as placeholder; the actual value should come from the metric columns
                return "0";
            }).ToList();

            double thresholdWan = massiveThreshold / 10000;
            sb.AppendLine($"var trace_amount = {{x:[{string.Join(",", allTimes)}], y:[{string.Join(",", amounts)}], mode:'lines', name:'Outside Amount', line:{{color:'blue',width:2}}, xaxis:'x2', yaxis:'y2'}};");

            // --- Subplot 3: Order Book Ratio ---
            var ratios = allData.Select(d =>
                d.BidAskRatio > 0 ? d.BidAskRatio.ToString("F2", CultureInfo.InvariantCulture) : "null"
            ).ToList();
            sb.AppendLine($"var trace_ratio = {{x:[{string.Join(",", allTimes)}], y:[{string.Join(",", ratios)}], mode:'lines', name:'Ask/Bid Ratio', line:{{color:'purple',width:2}}, xaxis:'x3', yaxis:'y3'}};");

            // Threshold value
            double obThreshold = 1.0;
            if (_config.ContainsKey("orderbook_bid_ask_ratio_min"))
                obThreshold = Convert.ToDouble(_config["orderbook_bid_ask_ratio_min"]);

            // Build all traces array
            sb.AppendLine("var allTraces = [trace_price, trace_dayhigh, trace_low0, trace_low1, trace_low2, trace_entry, trace_amount, trace_ratio].concat(trace_exits);");

            // Layout
            sb.AppendLine("var layout = {");
            sb.AppendLine($"  title: '{stockId} Early Entry Strategy (09:01-09:05) - {date}',");
            sb.AppendLine("  showlegend: true,");
            sb.AppendLine("  height: 900,");
            sb.AppendLine("  hovermode: 'x unified',");
            sb.AppendLine("  plot_bgcolor: 'white',");
            sb.AppendLine("  paper_bgcolor: 'white',");
            sb.AppendLine("  grid: {rows:3, columns:1, pattern:'independent', roworder:'top to bottom'},");
            sb.AppendLine("  xaxis: {showticklabels:false, showgrid:true, gridcolor:'lightgray'},");
            sb.AppendLine("  yaxis: {title:'Price', domain:[0.55,1.0]},");
            sb.AppendLine("  xaxis2: {showticklabels:false, showgrid:true, gridcolor:'lightgray'},");
            sb.AppendLine($"  yaxis2: {{title:'Amount (10k)', domain:[0.28,0.50]}},");
            sb.AppendLine("  xaxis3: {title:'Time', showgrid:true, gridcolor:'lightgray'},");
            sb.AppendLine($"  yaxis3: {{title:'Ratio', domain:[0.0,0.22]}},");
            // Add threshold shapes
            sb.AppendLine("  shapes: [");
            sb.AppendLine($"    {{type:'line', xref:'paper', x0:0, x1:1, yref:'y2', y0:{thresholdWan.ToString("F1", CultureInfo.InvariantCulture)}, y1:{thresholdWan.ToString("F1", CultureInfo.InvariantCulture)}, line:{{color:'red',width:2,dash:'dash'}}}},");
            sb.AppendLine($"    {{type:'line', xref:'paper', x0:0, x1:1, yref:'y3', y0:{obThreshold.ToString("F1", CultureInfo.InvariantCulture)}, y1:{obThreshold.ToString("F1", CultureInfo.InvariantCulture)}, line:{{color:'red',width:2,dash:'dash'}}}}");
            sb.AppendLine("  ]");
            sb.AppendLine("};");

            sb.AppendLine("Plotly.newPlot('chart', allTraces, layout, {displayModeBar:true, scrollZoom:true});");
        }

        /// <summary>
        /// Formats a time value for JavaScript string representation.
        /// </summary>
        private static string FormatTime(object timeValue)
        {
            if (timeValue is DateTime dt)
                return dt.ToString("yyyy-MM-dd HH:mm:ss.fff", CultureInfo.InvariantCulture);
            return timeValue?.ToString() ?? "";
        }
    }
}
